chiselFrontend/src/main/scala/chisel3/core/MultiClock.scala

```diff
@@ -0,0 +1,25 @@
+// See LICENSE for license details.
+
+package chisel3.core
+
+import scala.language.experimental.macros
+
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+import chisel3.internal.sourceinfo.{SourceInfo}
+
+private[chisel3] final case class ClockAndReset(clock: Clock, reset: Bool)
+
+object withClockAndReset {  // scalastyle:ignore object.name
+  def apply[T](clock: Clock, reset: Bool)(block: => T)(implicit sourceInfo: SourceInfo): T = {
```

Comment:
> To my knowledge, there currently isn't any way to access the clock or reset of the current Module inside of a function. Obviously this PR provides much more power for changing them so people are more likely to do so, just mentioning that this isn't removing anything that existed

This is where the horrendous (in my opinion) `AsyncScope` hack came from... you instantiate an AsyncScope then ask it what its clock is. This creates an empty module for no good reason.

But yeah. I'd prefer that if people really need to know the clock inside their function, then it should be passed as a parameter to the function.